File: 277. Find the Celebrity.cpp
// Forward declaration of the knows API.
bool knows(int a, int b);

class Solution {
public:
    int findCelebrity(int n) {
        int x = 0;
        for (int i = 1; i < n; i++) {
            if (knows(x, i)) {
                x = i;
            }
        }
        for (int i = 0; i < x; i++) {
            if (knows(x, i)) {
                return -1;
            }
        }
        for (int i = 0; i < n; i++) {
            if (!knows(i, x)) {
                return -1;
            }
        }
        return x;
    }
};

File: java/277. Find the Celebrity.java
/* The knows API is defined in the parent class Relation.
      boolean knows(int a, int b); */

// O(n)
public class Solution extends Relation {
    public int findCelebrity(int n) {
        int target = 0;
        for (int i = 1; i < n; ++i) {
            if (knows(target, i)) {
                target = i;
            }
        }
        for (int i = 0; i < n; ++i) {
            if (i != target && (!knows(i, target) || knows(target, i))) {
                return -1;
            }
        }
        return target;
    }
}
